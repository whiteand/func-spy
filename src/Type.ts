export enum Type {
  Undefined = 1 << 0,
  Null = 1 << 1,
  Boolean = 1 << 2,
  Number = 1 << 3,
  String = 1 << 4,
  Array = 1 << 5,
  Object = 1 << 6,
  Function = 1 << 7,
  Symbol = 1 << 8,
  Bigint = 1 << 9,
  Date = 1 << 10,
}
