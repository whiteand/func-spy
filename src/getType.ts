import { Type } from './Type';

export function getType(value: unknown): Type {
  if (value === null) return Type.Null;
  if (value === undefined) return Type.Undefined;
  switch (typeof value) {
    case 'number':
      return Type.Number;
    case 'boolean':
      return Type.Boolean;
    case 'function':
      return Type.Function;
    case 'string':
      return Type.String;
    case 'object':
      if (Array.isArray(value)) return Type.Array;
      if (value instanceof Date) return Type.Date;
      return Type.Object;
    default:
      throw new Error('cannot find typeof value');
  }
}
