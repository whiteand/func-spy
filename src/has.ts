import { Path } from './types';

export function has<K extends Path>(
  obj: unknown,
  k: K
): obj is Record<K, unknown> {
  if (!obj) return false;
  return Object.prototype.hasOwnProperty.call(obj, k);
}
