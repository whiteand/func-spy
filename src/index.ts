export * from './getMetadata';
export * from './getSpiedStruct';
export * from './getType';
export * from './has';
export * from './mergeMetadata';
export * from './spy';
export * from './Type';
export * from './types';
export * from './utils';
