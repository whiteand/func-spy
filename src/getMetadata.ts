import { getType } from './getType';
import { Type } from './Type';
import { IMetadata, Bitmask } from './types';

export function getMetadata<V>(arg: V): IMetadata<V> {
  return {
    children: Object.create(null) as IMetadata<V>['children'],
    data: [arg],
    path: [],
    type: getType(arg) as Bitmask<Type>,
  };
}
