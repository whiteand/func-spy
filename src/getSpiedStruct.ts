import { getType } from './getType';
import { Type } from './Type';
import { Path } from './types';

export function getSpiedStruct<T extends object>(
  obj: T,
  onMeet: (value: unknown, path: Path[]) => void
): T {
  const proxy: T = new Proxy<T>(obj, {
    get(t: T, k: string | symbol | number) {
      const child = t[k as keyof T];
      const childT = getType(child);
      onMeet(child, [k]);
      if (childT & (Type.Object | Type.Array)) {
        return getSpiedStruct(
          (child as unknown) as object,
          (subChild, subPath) => {
            onMeet(subChild, [k, ...subPath]);
          }
        );
      }
      if (childT & Type.Function) {
        return ((child as unknown) as Function).bind(proxy);
      }
      return child;
    },
  });
  return proxy;
}
