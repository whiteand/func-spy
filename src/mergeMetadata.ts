import { has } from './has';
import { Type } from './Type';
import { Bitmask, IMetadata } from './types';

export function mergeMetadata<T>(
  prev: IMetadata<T>,
  cur: IMetadata<T>
): IMetadata<T> {
  const newChildren = Object.fromEntries(
    Object.entries(prev.children).map(([k, m]) => {
      if (has(cur.children, k)) {
        const newChildM: typeof m = (cur.children as any)[k];
        return [
          k,
          mergeMetadata(
            m as IMetadata<unknown>,
            newChildM as IMetadata<unknown>
          ),
        ];
      }
      return [k, m];
    })
  );
  for (const [k, m] of Object.entries(cur.children)) {
    if (has(newChildren, k)) continue;
    (newChildren as any)[k] = m;
  }
  const newData = [...prev.data, ...cur.data];
  const newType = prev.type | cur.type;
  return {
    children: newChildren as any,
    data: newData,
    path: prev.path,
    type: newType as Bitmask<Type>,
  };
}
