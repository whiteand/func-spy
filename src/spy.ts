import { getMetadata } from './getMetadata';
import { getSpiedStruct } from './getSpiedStruct';
import { getType } from './getType';
import { has } from './has';
import { mergeMetadata } from './mergeMetadata';
import { Type } from './Type';
import { IMetadata, ISpyResult, Path } from './types';

export function spy<T, U>(fn: (arg: T) => U, arg: T): ISpyResult<T, U> {
  const argT = getType(arg);
  const rootMetadata = getMetadata(arg);
  if ((argT & (Type.Object | Type.Array)) === 0) {
    const res = fn(arg);
    return {
      result: res,
      metadata: rootMetadata,
    };
  }
  const onMeet = (child: unknown, path: Path[]) => {
    let parent: any = rootMetadata;
    for (let i = 0; i < path.length - 1; i++) {
      parent = parent.children[path[i]];
    }
    const lastKey = path[path.length - 1];

    const childMetadata = getMetadata(child);
    childMetadata.path = path;
    parent.children[lastKey] = has(parent.children, lastKey)
      ? mergeMetadata(
          parent.children[lastKey] as IMetadata<unknown>,
          childMetadata
        )
      : childMetadata;
  };
  type ArgType = T extends object ? T : never;

  const spiedStruct = getSpiedStruct<ArgType>(arg as ArgType, onMeet);
  const r = fn(spiedStruct);
  return {
    result: r,
    metadata: rootMetadata,
  };
}
