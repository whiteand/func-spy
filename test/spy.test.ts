import { IMetadata, iterate, Path, spy, Type, typeToStringTypes } from '../src';
const ID_TEST = /^[a-zA-Z_][a-zA-Z_0-9]*$/;
const NUM_TEST = /^\d+\.?\d*$/;
function pathToString(path: Path[]) {
  if (path.length <= 0) return '';
  const items = [
    typeof path[0] === 'number' || NUM_TEST.test(path[0].toString())
      ? `[${path[0].toString()}]`
      : ID_TEST.test(path[0].toString())
      ? path[0].toString()
      : `[${JSON.stringify(path[0].toString())}]`,
  ];
  for (let i = 1; i < path.length; i++) {
    const p = path[i];
    if (typeof p === 'number' || NUM_TEST.test(p.toString())) {
      items.push(`[${p.toString()}]`);
      continue;
    }
    const k = p.toString();
    if (ID_TEST.test(k)) {
      items.push(`.${k}`);
      continue;
    }
    items.push(`[${JSON.stringify(k)}]`);
  }
  return items.join('');
}
function collect(value: IMetadata<any>): string {
  const lines: string[] = [];
  iterate(m => {
    lines.push(
      `${pathToString(m.path)}: ${
        (m.type & Type.Function) === Type.Function
          ? m.data[0]?.name?.toString?.() + '()'
          : typeToStringTypes(m.type).join(' | ')
      }`
    );
  }, value);
  return lines.join(';\n');
}

describe('spy', () => {
  it('00. primitive', () => {
    const res = spy((x: number) => x * 2, 2);
    expect(res).toMatchSnapshot();
    expect(collect(res.metadata)).toMatchInlineSnapshot(`": number"`);
  });
  it('01. date', () => {
    const res = spy((x: Date) => x.toString(), new Date(2021, 11, 19));
    expect(res).toMatchSnapshot();
    expect(collect(res.metadata)).toMatchInlineSnapshot(`": date"`);
  });
  it('02. array', () => {
    const res = spy((x: number[]) => x.map(y => y * 2), [1, 2]);
    expect(res).toMatchSnapshot();
    expect(collect(res.metadata)).toMatchInlineSnapshot(`
      ": array;
      constructor: Array();
      length: number;
      map: map();
      [1]: number;
      [0]: number"
    `);
  });
  it('03. object', () => {
    const res = spy((x: Record<string, string>) => Object.keys(x), {
      andrew: 'beletskiy',
      vasilina: 'beletskaya',
    });
    expect(res).toMatchSnapshot();
    expect(collect(res.metadata)).toMatchInlineSnapshot(`": object"`);
  });
  it('04. nested', () => {
    const res = spy((x: any) => x?.data?.[0]?.name, {
      data: [{ name: 42 }],
    });
    expect(res).toMatchSnapshot();
    expect(collect(res.metadata)).toMatchInlineSnapshot(`
      ": object;
      data: array;
      data[0]: object;
      data[0].name: number"
    `);
  });
  test('05. docs', () => {
    const { result, metadata } = spy(state => state?.users?.[0]?.name, {
      users: [{ name: 'andrew' }],
    });

    expect(result).toMatchSnapshot();
    const res: any[] = [];
    iterate(node => {
      res.push([node.path], typeToStringTypes(node.type));
    }, metadata);
    expect(res).toMatchSnapshot();
  });
});
