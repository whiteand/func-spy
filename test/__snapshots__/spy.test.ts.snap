// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spy 00. primitive 1`] = `
Object {
  "metadata": Object {
    "children": Object {},
    "data": Array [
      2,
    ],
    "path": Array [],
    "type": 8,
  },
  "result": 4,
}
`;

exports[`spy 01. date 1`] = `
Object {
  "metadata": Object {
    "children": Object {},
    "data": Array [
      2021-12-18T22:00:00.000Z,
    ],
    "path": Array [],
    "type": 1024,
  },
  "result": "Sun Dec 19 2021 00:00:00 GMT+0200 (Eastern European Standard Time)",
}
`;

exports[`spy 02. array 1`] = `
Object {
  "metadata": Object {
    "children": Object {
      "0": Object {
        "children": Object {},
        "data": Array [
          1,
        ],
        "path": Array [
          "0",
        ],
        "type": 8,
      },
      "1": Object {
        "children": Object {},
        "data": Array [
          2,
        ],
        "path": Array [
          "1",
        ],
        "type": 8,
      },
      "constructor": Object {
        "children": Object {},
        "data": Array [
          [Function],
        ],
        "path": Array [
          "constructor",
        ],
        "type": 128,
      },
      "length": Object {
        "children": Object {},
        "data": Array [
          2,
        ],
        "path": Array [
          "length",
        ],
        "type": 8,
      },
      "map": Object {
        "children": Object {},
        "data": Array [
          [Function],
        ],
        "path": Array [
          "map",
        ],
        "type": 128,
      },
    },
    "data": Array [
      Array [
        1,
        2,
      ],
    ],
    "path": Array [],
    "type": 32,
  },
  "result": Array [
    2,
    4,
  ],
}
`;

exports[`spy 03. object 1`] = `
Object {
  "metadata": Object {
    "children": Object {},
    "data": Array [
      Object {
        "andrew": "beletskiy",
        "vasilina": "beletskaya",
      },
    ],
    "path": Array [],
    "type": 64,
  },
  "result": Array [
    "andrew",
    "vasilina",
  ],
}
`;

exports[`spy 04. nested 1`] = `
Object {
  "metadata": Object {
    "children": Object {
      "data": Object {
        "children": Object {
          "0": Object {
            "children": Object {
              "name": Object {
                "children": Object {},
                "data": Array [
                  42,
                ],
                "path": Array [
                  "data",
                  "0",
                  "name",
                ],
                "type": 8,
              },
            },
            "data": Array [
              Object {
                "name": 42,
              },
            ],
            "path": Array [
              "data",
              "0",
            ],
            "type": 64,
          },
        },
        "data": Array [
          Array [
            Object {
              "name": 42,
            },
          ],
        ],
        "path": Array [
          "data",
        ],
        "type": 32,
      },
    },
    "data": Array [
      Object {
        "data": Array [
          Object {
            "name": 42,
          },
        ],
      },
    ],
    "path": Array [],
    "type": 64,
  },
  "result": 42,
}
`;

exports[`spy 05. docs 1`] = `"andrew"`;

exports[`spy 05. docs 2`] = `
Array [
  Array [
    Array [],
  ],
  Array [
    "object",
  ],
  Array [
    Array [
      "users",
    ],
  ],
  Array [
    "array",
  ],
  Array [
    Array [
      "users",
      "0",
    ],
  ],
  Array [
    "object",
  ],
  Array [
    Array [
      "users",
      "0",
      "name",
    ],
  ],
  Array [
    "string",
  ],
]
`;
